cabal-install-solver/src/Distribution/Solver/Modular/Message.hs

```diff
@@ -206,12 +208,81 @@ data MergedPackageConflict = MergedPackageConflict {
   , versionConflict :: Maybe VR
   }
 
-showQPNPOpt :: QPN -> POption -> String
-showQPNPOpt qpn@(Q _pp pn) (POption i linkedTo) =
+data ProgressAction =
+    Trying
+  | Skipping
+  | Rejecting
+
+instance Show ProgressAction where
+  show Trying = "trying: "
+  show Skipping = "skipping: "
+  show Rejecting = "rejecting: "
+
+showQFNBool :: ProgressAction -> QFN -> Bool -> String
+showQFNBool a q b = show a ++ Flag.showQFNBool q b
+
+showQSNBool :: ProgressAction -> QSN -> Bool -> String
+showQSNBool a q b = show a ++ Flag.showQSNBool q b
+
+showOptions :: ProgressAction -> QPN -> [POption] -> String
+showOptions a q [p] = show a ++ showOption q p
+showOptions a q ps = show a ++ showIsOrVs q (tryVs ps)
+
+showOption :: QPN -> POption -> String
+showOption qpn@(Q _pp pn) (POption i linkedTo) =
   case linkedTo of
     Nothing  -> showPI (PI qpn i) -- Consistent with prior to POption
     Just pp' -> showQPN qpn ++ "~>" ++ showPI (PI (Q pp' pn) i)
 
+-- | A list of versions, or a list of instances.
+data IsOrVs = Is [POption] | Vs [Ver] deriving Show
+
+-- | Try to convert a list of options to a list of versions, or a list of
+-- instances if any of the options is linked (installed).  Singleton lists or
+-- empty lists are always converted to Is.
+-- >>> tryVs [v0, v1]
+-- Vs [mkVersion [0],mkVersion [1]]
+-- >>> tryVs [v0]
+-- Is [POption (I (mkVersion [0]) InRepo) Nothing]
+-- >>> tryVs [i0, i1]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-1-inplace"))) Nothing]
+-- >>> tryVs [i0, v1]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing,POption (I (mkVersion [1]) InRepo) Nothing]
+-- >>> tryVs [v0, i1]
+-- Is [POption (I (mkVersion [0]) InRepo) Nothing,POption (I (mkVersion [1]) (Inst (UnitId "foo-1-inplace"))) Nothing]
+-- >>> tryVs [i0]
+-- Is [POption (I (mkVersion [0]) (Inst (UnitId "foo-0-inplace"))) Nothing]
+-- >>> tryVs []
+-- Is []
+tryVs :: [POption] -> IsOrVs
+tryVs xs@[] = Is xs
+tryVs xs@[_] = Is xs
+tryVs xs
+    | any (\(POption (instI -> b0) (isJust -> b1)) -> b0 || b1) xs = Is xs
+    | otherwise =
+      let (vs, is) = L.partition ((== InRepo) . snd) [(v, l) | POption i _ <- xs, let I v l = i]
+      in if null is then Vs (fst `map` vs) else Is xs
```

Comment:
@grayjay do you know of a test that already uses installed packages? I've seen these listed in stackage `cabal.config` files, such as [stackage/lts-22.4](https://www.stackage.org/lts-22.4/cabal.config) without version numbers:

```
array installed,
base installed,
binary installed,
Cabal installed,
Cabal-syntax installed,
containers installed,
deepseq installed,
...
Win32 installed,
xhtml installed,
```

In this case are you expecting the output to look something like:

`rejecting: base; x.y.z, x.y.z, installed, x.y.z, x.y.z`

Reply:
I've added the following unit tests that test skipping:

```
$ cabal run cabal-install:unit-tests -- -p "5960"
...
Unit Tests
  UnitTests.Distribution.Solver.Modular.Solver
    Solver log
      package versions abbreviation (issue #5960)
        show skipping singleton:                OK
        show skipping singleton, installed:     OK
        show skipping versions list:            OK
        show skipping versions list, installed: OK

All 4 tests passed (0.00s)
```

I couldn't see how to trigger multiple rejections like in the pandoc example of #9559.

Reply:
> I don't think that it's necessary to exclude installed packages from the new formatting, if you are interested in handling them, too.

We shorten the rejecting and skipping messages a lot by putting the package name in front of a list of versions and not repeating it hyphenated with the versions. I tried the same thing with installed packages and thought it harder to read and not so effective at shortening. Here is the test example I tried:

```
[__0] trying: B-1.0.0 (user goal)
[__1] next goal: A (dependency of B)
[__1] rejecting: A; 3.0.0/installed-3.0.0 (conflict: B => A==4.0.0)
[__1] skipping: A; 2.0.0/installed-2.0.0, 1.0.0/installed-1.0.0 ...
[__1] fail (backjumping, conflict set: A, B)
```

I would rather not extend this pull request to include abbreviating lists that include installed packages.

Reply:
@grayjay should I bring back the `PackageTests/ShortRejections` tests? They were able to trigger multiple rejections like in the pandoc example of #9559.

Reply:
Thanks for adding the new unit tests. I think that those are enough, and this change doesn't need the package test. I'm guessing that packages are rejected rather than skipped when there is a conflict due to a global constraint.